crates/turbopack-ecmascript/src/references/require_context.rs

```diff
@@ -0,0 +1,267 @@
+    create_visitor,
+    references::{
+        pattern_mapping::{PatternMappingVc, ResolveType::Cjs},
+        AstPathVc,
+    },
+    resolve::{cjs_resolve, try_to_severity},
+    CodeGenerateableVc,
+};
+
+#[turbo_tasks::value(transparent)]
+pub(crate) struct FlatDirList(Vec<(String, FileSystemPathVc)>);
+
+#[turbo_tasks::function]
+pub(crate) async fn list_dir(
+    dir: FileSystemPathVc,
+    recursive: bool,
+    filter: RegexVc,
+) -> Result<FlatDirListVc> {
+    let root = &*dir.await?;
+    let filter = &*filter.await?;
+
+    let mut queue = VecDeque::with_capacity(1);
+    queue.push_back(dir);
+
+    let mut list = Vec::new();
+
+    while let Some(dir) = queue.pop_front() {
+        let dir_content = dir.read_dir().await?;
+        let entries = match &*dir_content {
+            DirectoryContent::Entries(entries) => entries,
+            DirectoryContent::NotFound => continue,
+        };
+
+        for (_, entry) in entries {
+            match entry {
+                DirectoryEntry::File(path) => {
+                    if let Some(relative_path) = root.get_relative_path_to(&*path.await?) {
+                        if filter.is_match(&relative_path) {
+                            list.push((relative_path, *path));
+                        }
+                    }
+                }
+                DirectoryEntry::Directory(path) if recursive => {
+                    queue.push_back(*path);
+                }
+                // ignore everything else
+                _ => {}
+            }
+        }
+    }
+
+    Ok(FlatDirListVc::cell(list))
+}
+
+#[turbo_tasks::value]
+#[derive(Debug)]
+pub(crate) struct RequireContextMapEntry {
+    pub origin_relative: String,
+    pub request: RequestVc,
+    pub result: ResolveResultVc,
+}
+
+#[turbo_tasks::value(transparent)]
+pub(crate) struct RequireContextMap(IndexMap<String, RequireContextMapEntry>);
+
+#[turbo_tasks::function]
+pub(crate) async fn generate_require_context_map(
+    origin: ResolveOriginVc,
+    dir: FileSystemPathVc,
+    recursive: bool,
+    filter: RegexVc,
+    issue_source: OptionIssueSourceVc,
+    issue_severity: IssueSeverityVc,
+) -> Result<RequireContextMapVc> {
+    let origin_path = &*origin.origin_path().parent().await?;
+
+    let list = &*list_dir(dir, recursive, filter).await?;
+
+    let mut map = IndexMap::new();
+
+    for (context_relative, path) in list {
+        if let Some(origin_relative) = origin_path.get_relative_path_to(&*path.await?) {
+            let request = RequestVc::parse(Value::new(origin_relative.clone().into()));
+            let result = cjs_resolve(origin, request, issue_source, issue_severity);
+
+            map.insert(
+                context_relative.clone(),
+                RequireContextMapEntry {
+                    origin_relative,
+                    request,
+                    result,
+                },
+            );
+        } else {
+            // TODO: does this need to throw an error?
```

Comment:
Is this even possible? If it's theoretically impossible, we should throw an invariant error here at least.